### PR TITLE
[Issue 8300][Java client] Support cancelling message & batch futures returned from Reader & Consumer

### LINKED ISSUE #8300: [Java Client] Support cancelling message & batch futures returned by Reader and Consumer APIs
**Problem description**

Currently, it's not possible to cancel the message & batch futures returned by Reader and Consumer APIs. 
This leads to several problems when the Reader or Consumer instance is long-living. If no messages are received,
the futures will be kept in memory until messages are received. This causes memory leaks. In addition, the returned
message would get consumed by an already cancelled future which the application using Pulsar has already discarded, for example as a result of a timeout from the application's perspective

There is a gap in the asynchronous Java API compared to the synchronous API. In the synchronous API, it's possible to set a timeout for Reader.readNext or Consumer.receive . **When the future can be cancelled, it also makes it possible to have timeouts for asynchronous calls without making any API changes**. In JDK9+, there is [".orTimeout" method on CompletableFuture](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CompletableFuture.html#orTimeout(long,java.util.concurrent.TimeUnit)). 

For example, after adding cancellation & timeout support, it will be possible to read the next message with a timeout in this way: 
```java
CompletableFuture<Message<?>> result = reader.readNextAsync().orTimeout(1, TimeUnit.SECONDS);
```

**Solution**
The futures returned by `org.apache.pulsar.client.api.Reader.readNextAsync`, `org.apache.pulsar.client.api.Consumer.receiveAsync` and `org.apache.pulsar.client.api.Consumer.batchReceiveAsync` methods should unregister the futures from internal datastructures when the future gets cancelled or it gets completed exceptionally with TimeoutException.

### PR DESCRIPTION
**Pull request by lhotari**
Fixes #8300, #8307

### Motivation

Explained in #8300

### Modifications

The futures returned by `org.apache.pulsar.client.api.Reader.readNextAsync`, `org.apache.pulsar.client.api.Consumer.receiveAsync` and `org.apache.pulsar.client.api.Consumer.batchReceiveAsync` methods 
return a `CompletableFuture` that contains a "whenComplete" callback which handles `CancellationException & TimeoutException` . The callback action will remove the registered futures from the internal "pendingReceives" and "pendingBatchReceives" futures.
The changes attempt to also against possible race conditions that could happen when cancellation and a message get received at the exactly same moment and the future removed from the queue has already been cancelled. 

While adding unit tests, some minor refactoring was made to start adding a ClientTestFixtures class that would make it easier to share some test mocks that are used in the client internal unit tests. The intention of this change was to start lowering the barrier for adding better test coverage for further changes. 

### Verifying this change

Unit tests have been added to verify that cancelling message & batch futures remove the future from the internal "pendingReceives" and "pendingBatchReceives" queues.

### Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java [hunk 1]
```diff
@@ -733,9 +821,11 @@ private void pendingBatchReceiveTask(Timeout timeout) throws Exception {
                 if (diff <= 0) {
                     // The diff is less than or equal to zero, meaning that the batch receive has been timed out.
                     // complete the OpBatchReceive and continue to check the next OpBatchReceive in pendingBatchReceives.
-                    OpBatchReceive<T> op = pendingBatchReceives.poll();
-                    completeOpBatchReceive(op);
-                    firstOpBatchReceive = pendingBatchReceives.peek();
+                    OpBatchReceive<T> op = pollNextBatchReceive();
+                    if (op != null) {
+                        completeOpBatchReceive(op);
+                    }
```
Comment (lhotari) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:827]: this seems to fix #8307

**Review by jiazhai — APPROVED** [review]
+1, Thanks for the great work.

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**